client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -43,17 +46,44 @@ impl AddrCache {
 			return;
 		}
 
+		addresses.sort_unstable_by(|a, b| a.as_ref().cmp(b.as_ref()));
+
 		// Insert into `self.peer_id_to_authority_id`.
 		let peer_ids = addresses.iter()
 			.map(|a| peer_id_from_multiaddr(a))
 			.filter_map(|peer_id| peer_id);
-		for peer_id in  peer_ids {
-			self.peer_id_to_authority_id.insert(peer_id, authority_id.clone());
+		for peer_id in peer_ids.clone() {
+			let former_auth = match self.peer_id_to_authority_id.insert(peer_id, authority_id.clone()) {
+				Some(a) if a != authority_id => a,
+				_ => continue,
+			};
+
+			// PeerId was associated to a different authority id before.
+			// Remove corresponding authority from `self.authority_id_to_addresses`.
+			let former_auth_addrs = match self.authority_id_to_addresses.get_mut(&former_auth) {
+				Some(a) => a,
+				None => { debug_assert!(false); continue }
+			};
+			former_auth_addrs.retain(|a| peer_id_from_multiaddr(a).map_or(false, |p| p != peer_id));
 		}
 
 		// Insert into `self.authority_id_to_addresses`.
-		addresses.sort_unstable_by(|a, b| a.as_ref().cmp(b.as_ref()));
-		self.authority_id_to_addresses.insert(authority_id, addresses);
+		for former_addr in
+			self.authority_id_to_addresses.insert(authority_id.clone(), addresses.clone()).unwrap_or_default()
+		{
+			// Must remove from `self.peer_id_to_authority_id` any PeerId formerly associated
+			// to that authority but that can't be found in its new addresses.
+
+			let peer_id = match peer_id_from_multiaddr(&former_addr) {
+				Some(p) => p,
+				None => { debug_assert!(false); continue },
```

Comment:
unrelated dumb question: why do we even keep addresses with no peer_ids in the addr_cache?
also not sure how to interpret `map_or(false` above?

Reply:
It is in my opinion better if the cache contains exactly the data that is present on the DHT, without any post-processing.

Reply:
> it is in my opinion better if the cache contains exactly the data that is present on the DHT

why do we remove it in
> map_or(false above?

Reply:
Changed the `map_or`

Reply:
but now the `debug_assert!` no longer holds? :sweat_smile: